[JSC] Make JSON.parse even faster with identifier-hint and putDirect fast path

Constellation opened the pull request:
#### 1f24101561da7ba92edc5b8efab210af2393c7a6
<pre>
[JSC] Make JSON.parse even faster with identifier-hint and putDirect fast path
<a href="https://bugs.webkit.org/show_bug.cgi?id=260921">https://bugs.webkit.org/show_bug.cgi?id=260921</a>
rdar://114715367

Reviewed by Keith Miller.

This patch accelerates JSON.parse even more by introducing two optimizations.

1. We add IndeitiferHint to JSON.parse&apos;s lexing fast path. This slightly changes how we scan the string,
   given the information that this will be used as an identifier for objects. We swap the ordering of
   comparison for UChar JSON.parse in isSafeStringCharacterForIdentifier, since identifier is, in almost all cases, ASCII!
2. We found that `putDirect` function call in JSON.parse fast path is costly, and if we can inline the fast cases of this,
   we can get performance boost. Given that this is JSON.parse&apos;s fast path for JSObject creation, we know that
    a. The object is definitely JSFinalObject.
    b. The object rarely has duplicate properties.
    c. Many same-shaped objects would be created from JSON. Thus very likely, there is already an existing Structure.
   Based on this observation, we add super fast path in JSON.parse which does existing Structure transition quickly.
3. Added JSON.parse microbenchmarks extracted from Speedometer3.0.

                                                     ToT                     Patched

todomvc-javascript-es5-json-parse              58.1822+-0.2031     ^     52.2885+-0.3449        ^ definitely 1.1127x faster
vanilla-es2015-babel-webpack-todomvc-json-parse
                                               57.4813+-0.5289     ^     54.7597+-0.1805        ^ definitely 1.0497x faster
todomvc-javascript-es6-webpack-json-parse
                                               89.5463+-0.5104     ^     83.3758+-0.5923        ^ definitely 1.0740x faster
vanilla-es2015-todomvc-json-parse              57.2404+-0.4405     ^     54.6702+-0.1448        ^ definitely 1.0470x faster
vanilla-todomvc-json-parse                     42.4967+-0.2446     ^     40.6605+-0.0723        ^ definitely 1.0452x faster

* JSTests/microbenchmarks/todomvc-javascript-es5-json-parse.js: Added.
(test):
* JSTests/microbenchmarks/todomvc-javascript-es6-webpack-json-parse.js: Added.
(test):
* Source/JavaScriptCore/runtime/JSObject.cpp:
(JSC::JSObject::putDirectForJSONSlow):
* Source/JavaScriptCore/runtime/JSObject.h:
* Source/JavaScriptCore/runtime/JSObjectInlines.h:
(JSC::JSObject::putDirectInternal):
* Source/JavaScriptCore/runtime/LiteralParser.cpp:
(JSC::isJSONWhiteSpace):
(JSC::LiteralParser&lt;CharType&gt;::Lexer::lex):
(JSC::LiteralParser&lt;CharType&gt;::Lexer::next):
(JSC::LiteralParser&lt;CharType&gt;::Lexer::nextMaybeIdentifier):
(JSC::isSafeStringCharacterForIdentifier):
(JSC::LiteralParser&lt;CharType&gt;::Lexer::lexString):
(JSC::LiteralParser&lt;CharType&gt;::parseRecursively):
* Source/JavaScriptCore/runtime/LiteralParser.h:
* Source/JavaScriptCore/runtime/StructureInlines.h:
(JSC::Structure::addPropertyTransitionToExistingStructureImpl):

Canonical link: <a href="https://commits.webkit.org/267507@main">https://commits.webkit.org/267507@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c68848125034d90c9280dd304fe4182727d95d2f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16801 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17124 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17578 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18583 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15738 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20357 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17263 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/18017 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17000 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/17369 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/14549 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19382 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/14613 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/15233 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/21982 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/14493 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/15602 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/15400 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/19716 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/16029 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16001 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/13573 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/18359 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15176 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/4292 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19540 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/19582 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2069 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15832 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/4143 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/runtime/LiteralParser.cpp:
```diff
@@ -1273,8 +1300,35 @@ JSValue LiteralParser<CharType>::parseRecursively(VM& vm, uint8_t* stackLimit)
+                    ASSERT(!object->getDirect(offset) || !JSValue::encode(object->getDirect(offset)));
+                    object->putDirectOffset(vm, offset, value);
+                    object->setStructure(vm, newStructure);
+                    ASSERT(!newStructure->mayBePrototype()); // There is no way to make it prototype object.
```
Is this true? What happens if you do:  ```
let o1 = JSON.parse("{ a: 1 }");
o2.__proto__ = o1;
JSON.parse("{ a: 1 }");```

Constellation replied:
Yes, this is true, since `mayBePrototype` bit gets moved to Structure, and setting `__proto__` causes structure transition now.

kmiller68 commented on Source/JavaScriptCore/runtime/LiteralParser.cpp:
```diff
@@ -703,10 +703,11 @@ static constexpr const bool safeStringLatin1CharactersInStrictJSON[256] = {
 static ALWAYS_INLINE bool isJSONWhiteSpace(const CharType& c)
 {
-    return isLatin1(c) && tokenTypesOfLatin1Characters[c] == TokErrorSpace;
+    return tokenTypesOfLatin1Characters[static_cast<uint8_t>(c)] == TokErrorSpace && isLatin1(c);
```
Why make this change?

Constellation replied:
Because this is almost always ASCII characters (otherwise, it becomes SyntaxError). So this is better ordering.

kmiller68 reviewed (approved):
r=me with comments

Constellation committed:
[JSC] Make JSON.parse even faster with identifier-hint and putDirect fast path
https://bugs.webkit.org/show_bug.cgi?id=260921
rdar://114715367

Reviewed by Keith Miller.

This patch accelerates JSON.parse even more by introducing two optimizations.

1. We add IndeitiferHint to JSON.parse's lexing fast path. This slightly changes how we scan the string,
   given the information that this will be used as an identifier for objects. We swap the ordering of
   comparison for UChar JSON.parse in isSafeStringCharacterForIdentifier, since identifier is, in almost all cases, ASCII!
2. We found that `putDirect` function call in JSON.parse fast path is costly, and if we can inline the fast cases of this,
   we can get performance boost. Given that this is JSON.parse's fast path for JSObject creation, we know that
    a. The object is definitely JSFinalObject.
    b. The object rarely has duplicate properties.
    c. Many same-shaped objects would be created from JSON. Thus very likely, there is already an existing Structure.
   Based on this observation, we add super fast path in JSON.parse which does existing Structure transition quickly.
3. Added JSON.parse microbenchmarks extracted from Speedometer3.0.

                                                     ToT                     Patched

todomvc-javascript-es5-json-parse              58.1822+-0.2031     ^     52.2885+-0.3449        ^ definitely 1.1127x faster
vanilla-es2015-babel-webpack-todomvc-json-parse
                                               57.4813+-0.5289     ^     54.7597+-0.1805        ^ definitely 1.0497x faster
todomvc-javascript-es6-webpack-json-parse
                                               89.5463+-0.5104     ^     83.3758+-0.5923        ^ definitely 1.0740x faster
vanilla-es2015-todomvc-json-parse              57.2404+-0.4405     ^     54.6702+-0.1448        ^ definitely 1.0470x faster
vanilla-todomvc-json-parse                     42.4967+-0.2446     ^     40.6605+-0.0723        ^ definitely 1.0452x faster

* JSTests/microbenchmarks/todomvc-javascript-es5-json-parse.js: Added.
(test):
* JSTests/microbenchmarks/todomvc-javascript-es6-webpack-json-parse.js: Added.
(test):
* Source/JavaScriptCore/runtime/JSObject.cpp:
(JSC::JSObject::putDirectForJSONSlow):
* Source/JavaScriptCore/runtime/JSObject.h:
* Source/JavaScriptCore/runtime/JSObjectInlines.h:
(JSC::JSObject::putDirectInternal):
* Source/JavaScriptCore/runtime/LiteralParser.cpp:
(JSC::isJSONWhiteSpace):
(JSC::LiteralParser<CharType>::Lexer::lex):
(JSC::LiteralParser<CharType>::Lexer::next):
(JSC::LiteralParser<CharType>::Lexer::nextMaybeIdentifier):
(JSC::isSafeStringCharacterForIdentifier):
(JSC::LiteralParser<CharType>::Lexer::lexString):
(JSC::LiteralParser<CharType>::parseRecursively):
* Source/JavaScriptCore/runtime/LiteralParser.h:
* Source/JavaScriptCore/runtime/StructureInlines.h:
(JSC::Structure::addPropertyTransitionToExistingStructureImpl):

Canonical link: https://commits.webkit.org/267507@main
